solr/core/src/java/org/apache/solr/search/KeepAliveRegenerator.java

```diff
@@ -0,0 +1,593 @@
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.FixedBitSet;
+import org.apache.lucene.util.RamUsageEstimator;
+import org.apache.lucene.util.automaton.ByteRunAutomaton;
+import org.apache.solr.common.MapWriter;
+import org.apache.solr.core.SolrConfig;
+import org.apache.solr.search.SolrCache.MetaEntry;
+import org.apache.solr.util.IOFunction;
+
+/**
+ * A regenerator capable of lazily warming entries where possible (passing stale entries along
+ * unmodified, with metadata to allow using the stale entry to reconstruct new entries on-demand,
+ * without actually needing to run the query again over segments that still exist in the index).
+ *
+ * <p>Cache entries with cross-doc/cross-segment dependencies (e.g., {@link JoinQuery}) cannot be
+ * warmed lazily, but this regenerator may still be configured to eagerly warm such queries.
+ *
+ * <p>Especially because this regenerator is designed to be used in conjunction with {@code
+ * autowarm="100%"}, it can be important to prevent unused queries from being carried forward
+ * indefinitely. Accordingly, this class extends {@link MetaCacheRegenerator}, wrapping entries in
+ * the associated cache to record (and update) each entry's last access time, allowing to drop (not
+ * warm) entries that haven't been accessed recently enough.
+ *
+ * <p>There are 4 configuration parameters (configured by attributes in the element configuring the
+ * associated cache):
+ *
+ * <ul>
+ *   <li>{@code regenKeepAlive}: max time since last access for cache entries to warm lazily
+ *   <li>{@code eagerKeepAlive}: max time since last access for cache entries to warm eagerly
+ *   <li>{@code preferLazy}: determines warming behavior for cache entries that are capable of being
+ *       warmed lazily, but that also meet the criteria configured for {@code eagerKeepAlive}. If
+ *       {@code false} (the default), such entries will be warmed eagerly (though still leveraging
+ *       the stale cache entry to make warming more efficient). If set to {@code "true"}, no eager
+ *       evaluation will be performed at warming time; the stale entry will be carried forward to
+ *       the new cache as-is, and used lazily to reconstruct a new cache entry on-demand, if needed.
+ *   <li>{@code overlapThreshold}: (default {@code 0.5}) determines the minimum segment overlap
+ *       threshold that allows cache entries to be carried forward lazily. Below this threshold, the
+ *       costs (heap consumed) are considered to outweigh the potential benefits of partial cache
+ *       entry reconstruction.
+ * </ul>
+ */
+public class KeepAliveRegenerator<M extends MetaEntry<Query, DocSet, M>>
+    extends MetaCacheRegenerator<Query, DocSet, M> {
+
+  private static final int DEFAULT_REGEN_KEEPALIVE_MINUTES = 2;
+  private static final long DEFAULT_REGEN_KEEPALIVE_NANOS =
+      TimeUnit.MINUTES.toNanos(DEFAULT_REGEN_KEEPALIVE_MINUTES);
+
+  private static final double DEFAULT_OVERLAP_THRESHOLD = 0.5;
```

Comment:
Curious of how likely would it drop below the threshold.

I assume this could only happen when:
1. The core is new, even a single new segment from commit could trigger it
2. Alot of merges happened? AFAIK, most merge policy should not modify > 50% of the entries? And that mature cores would unlikely to run into this?

So almost sounds like only newer cores could trigger this? Or am i missing some scenarios :) ?